File: app/src/main/java/com/foreverrafs/numericals/adapter/OperationsMenuAdapter.java
package com.foreverrafs.numericals.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.foreverrafs.numericals.R;
import com.foreverrafs.numericals.model.OperationMenu;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by Rafsanjani on 6/5/2019
 */
public class OperationsMenuAdapter extends RecyclerView.Adapter<OperationsMenuAdapter.MenuViewHolder> {
    private final List<OperationMenu> menuList;


    public OperationsMenuAdapter(List<OperationMenu> menuList) {
        this.menuList = menuList;
    }

    @NonNull
    @Override
    public MenuViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_menu, parent, false);

        return new MenuViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull MenuViewHolder holder, int position) {
        OperationMenu menu = menuList.get(position);

        holder.menuImage.setImageResource(menu.getImageResource());
        holder.menuTitle.setText(menu.getTitle());
    }

    @Override
    public int getItemCount() {
        return menuList.size();
    }

    class MenuViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.image_menu_item)
        ImageView menuImage;

        @BindView(R.id.text_menu_title)
        TextView menuTitle;

        MenuViewHolder(@NonNull View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }
}

File: app/src/main/java/com/foreverrafs/numericals/adapter/OperationsMenuAdapter.kt
package com.foreverrafs.numericals.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.foreverrafs.numericals.adapter.OperationsMenuAdapter.MenuViewHolder
import com.foreverrafs.numericals.databinding.ItemMenuBinding
import com.foreverrafs.numericals.model.OperationMenu

/** Created by Rafsanjani on 6/5/2019 */
class OperationsMenuAdapter(private val menuList: List<OperationMenu>) : RecyclerView.Adapter<MenuViewHolder>() {
    private lateinit var itemClickListener: MenuItemClickListenener

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MenuViewHolder {
        val binding = ItemMenuBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return MenuViewHolder(binding)
    }

    override fun onBindViewHolder(holder: MenuViewHolder, position: Int) {
        holder.bind(position)
    }

    fun setOnItemClickListenener(itemClickListenener: MenuItemClickListenener) {
        itemClickListener = itemClickListenener
    }

    override fun getItemCount(): Int {
        return menuList.size
    }

    fun interface MenuItemClickListenener {
        fun onMenuItemClicked(menuItemType: Int)
    }

    inner class MenuViewHolder(val binding: ItemMenuBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(position: Int) {
            val menu = menuList[position]
            binding.imageMenuItem.setImageResource(menu.imageResource)
            binding.textMenuTitle.text = menu.title
            //only attach a listener if there are sub menus

            binding.imageMenuItem.setOnClickListener {
                val menuCategory = menu.menuCategory
                itemClickListener.onMenuItemClicked(menuCategory)
            }
        }
    }

}